File: app-play/app/assets/javascripts/dashboard/CashAccountController.js
/**
 * Cash Account Controller
 * @author lawrence.daniels@gmail.com
 */
(function() {
    var app = angular.module('shocktrade');

    /**
     * Cash Account Controller
     * @author lawrence.daniels@gmail.com
     */
    app.controller('CashAccountController', ['$scope', '$log', 'toaster', 'MySession',
        function ($scope, $log, toaster, MySession) {

            /////////////////////////////////////////////////////////////////////
            //          Cash Account Functions
            /////////////////////////////////////////////////////////////////////

            $scope.asOfDate = function () {
                var participant = MySession.getParticipant();
                return participant.lastTradeTime || new Date();
            };

            $scope.getTotalOrders = function () {
                return $scope.getTotalBuyOrders() + $scope.getTotalSellOrders();
            };

            $scope.getTotalEquity = function () {
                return $scope.getTotalInvestment() + MySession.getFundsAvailable();
            };

            $scope.getTotalInvestment = function () {
                var total = 0;
                angular.forEach($scope.getPositions(), function (p) {
                    total += p.netValue;
                });
                return total;
            };

            $scope.getTotalBuyOrders = function () {
                var participant = MySession.getParticipant();
                var total = 0;
                if (participant != null) {
                    angular.forEach(participant.orders, function (o) {
                        if (o.orderType == 'BUY') {
                            total += o.price * o.quantity + o.commission;
                        }
                    });
                }
                return total;
            };

            $scope.getTotalSellOrders = function () {
                var participant = MySession.getParticipant();
                var total = 0;
                if (participant != null) {
                    angular.forEach(participant.orders, function (o) {
                        if (o.orderType == 'SELL') {
                            total += o.price * o.quantity + o.commission;
                        }
                    });
                }
                return total;
            };

        }]);

})();

File: app-js/src/main/scala/com/shocktrade/javascript/dashboard/CashAccountController.scala
package com.shocktrade.javascript.dashboard

import org.scalajs.angularjs.toaster.Toaster
import org.scalajs.angularjs.{Controller, Scope, injected}
import com.shocktrade.javascript.MySessionService

import scala.language.postfixOps
import scala.scalajs.js

/**
  * Cash Account Controller
  * @author lawrence.daniels@gmail.com
  */
class CashAccountController($scope: CashAccountScope, toaster: Toaster,
                            @injected("MySessionService") mySession: MySessionService)
  extends Controller {

  /////////////////////////////////////////////////////////////////////
  //          Public Functions
  /////////////////////////////////////////////////////////////////////

  $scope.asOfDate = () => mySession.cashAccount_?.flatMap(a => Option(a.asOfDate)) getOrElse new js.Date()

  $scope.getTotalOrders = () => Seq("BUY", "SELL") map computeTotalOrdersByType sum

  $scope.getTotalEquity = () => $scope.getTotalInvestment() + $scope.getFundsAvailable()

  $scope.getTotalBuyOrders = () => computeTotalOrdersByType(orderType = "BUY")

  $scope.getTotalSellOrders = () => computeTotalOrdersByType(orderType = "SELL")

  $scope.getFundsAvailable = () => mySession.cashAccount_?.map(_.cashFunds) getOrElse 0d

  $scope.getTotalInvestment = () => {
    mySession.participant.map(_.positions.filter(_.accountType == "CASH")).map(_ map (_.netValue) sum) getOrElse 0d
  }

  /////////////////////////////////////////////////////////////////////
  //          Private Functions
  /////////////////////////////////////////////////////////////////////

  private def computeTotalOrdersByType(orderType: String) = {
    var total = 0d
    mySession.participant foreach (_.orders filter (o => o.orderType == orderType && o.accountType == "CASH") foreach { o =>
      total += o.price * o.quantity + o.commission
    })
    total
  }

}

/**
  * Cash Account Controller Scope
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait CashAccountScope extends Scope {
  var asOfDate: js.Function0[js.Date] = js.native
  var getTotalOrders: js.Function0[Double] = js.native
  var getTotalEquity: js.Function0[Double] = js.native
  var getTotalBuyOrders: js.Function0[Double] = js.native
  var getTotalSellOrders: js.Function0[Double] = js.native
  var getFundsAvailable: js.Function0[Double] = js.native
  var getTotalInvestment: js.Function0[Double] = js.native

}